SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityPendingTaskQueueProcessor.java

```diff
@@ -0,0 +1,419 @@
+        pendingTaskId.getId(),
+        System.currentTimeMillis() - start
+      );
+      // Run this on the single thread responsible for offer matching and launches
+      CompletableFuture
+        .runAsync(() -> matchAndLaunch(start, pendingTaskId), taskLaunchExecutor)
+        .join();
+    } catch (Exception e) {
+      LOG.error("Exception while launching task {}", pendingTaskId, e);
+    } finally {
+      lock.unlock(
+        pendingTaskId.getRequestId(),
+        LOCK_NAME,
+        Optional
+          .ofNullable(lockStarts.remove(pendingTaskId.getId()))
+          .orElse(System.currentTimeMillis())
+      );
+    }
+  }
+
+  // Everything in this method is run behind a request-level lock and offer lock
+  private void matchAndLaunch(long start, SingularityPendingTaskId pendingTaskId) {
+    try {
+      long startLaunch = System.currentTimeMillis();
+      try {
+        Optional<SingularityPendingTask> maybePendingTask = taskManager.getPendingTask(
+          pendingTaskId
+        );
+        if (!maybePendingTask.isPresent()) {
+          LOG.error("Could not find pending task for id {}", pendingTaskId);
+          return;
+        }
+        SingularityPendingTask toLaunch = maybePendingTask.get();
+        SingularityTaskRequestHolder taskRequestHolder = new SingularityTaskRequestHolder(
+          taskRequestManager.getTaskRequest(toLaunch),
+          defaultResources,
+          defaultCustomExecutorResources
+        );
+        List<SingularityOfferHolder> offers = offerManager.checkoutOffers();
+        if (offers.isEmpty()) {
+          preLockQueue.put(toLaunch.getPendingTaskId());
+          return;
+        }
+        LOG.info("Checking {} offers for pending task {}", offers.size(), pendingTaskId);
+        SingularityOfferHolder bestOffer = null;
+        double maxOfferScore = 0.0;
+        for (SingularityOfferHolder offer : offers) {
+          double offerScore = scoreOffer(taskRequestHolder, offer);
+          if (offerScore > maxOfferScore) {
+            bestOffer = offer;
+          }
+          if (maxOfferScore >= mesosConfiguration.getGoodEnoughScoreThreshold()) {
+            break;
+          }
+        }
+        if (bestOffer != null) {
+          acceptTask(bestOffer, taskRequestHolder);
+          List<Offer> usedOffers = bestOffer.getOffers();
+          List<Offer> unused = bestOffer.launchTasksAndGetUnusedOffers(schedulerClient);
+          usedOffers.removeAll(unused);
+          usedOffers.stream().map(Offer::getId).forEach(offerManager::useOffer);
+          unused.stream().map(Offer::getId).forEach(offerManager::returnOffer);
+
+          synchronized (handled) {
+            handled.remove(pendingTaskId);
+          }
+        } else {
+          // Back on the queue and try again. Temporarily free up lock for other pollers first
+          preLockQueue.put(toLaunch.getPendingTaskId());
+        }
+        offers.remove(bestOffer);
```

Comment:
Is it possible to match a big enough offer to multiple tasks? If so, is there a way to safely subtract resources from the offer and return it?